### ISSUE TITLE
Make it easier to override package implementations

### ISSUE THREAD

**Issue by dalcde** [plastex/plastex#167]
When people use `plasTeX` for complicated projects, it is inevitable that some existing package implementations will be unsuitable for them. It should be easy for users to override the package definitions.

Currently, packages are imported by attempting to import the python module with the same name as the package, and this works since `Base/LaTeX/Packages.py` adds `plasTeX/Packages/` to `sys.path`.

Under this system, to override a package implementation, one has to put their `package.py` file in a directory that comes earlier than `plasTeX/Packages/` in `sys.path`, such as the directory of the running script. If one wants to use a custom directory, they must carefully add the directory to `sys.path` before `plasTeX.Base` gets (transitively) imported (or use `sys.path.insert(.., 0)`).

In theory, there is also the possibility that there is a latex package with no python implementation, but there happens to be a python module with the same path.

I propose that we introduce a new array in `config` for places to look for python implementations. If the package is not found there, then we check `cwd,`, then `plasTeX/Packages`. The latter are made special cases instead of part of the config, because these paths are determined at runtime.

We should also rewrite code to let us override implementations in `Base` as well, which afaik cannot currently be done. One way to do so is that in the config, we should specify directories which contain (optional) subdirectories `Packages/`, `Base/{LaTeX/TeX}`.

**Comment by PatrickMassot** [comment]
I agree the current "solutions" are way too brittle and clunky. The situation with Packages is probably rather easy to fix by modifying `Base/LaTeX/Packages.py`. Overriding implementations in `Base` seems much harder.

Please also keep in mind that I'd like to have some plugin mechanism, as discussed in #159.

**Comment by dalcde** [comment]
I think using a path-like object to look for package implementations is
reasonable. After all, that's how TeX does it.

I think a cleaner approach would be to have a list pre-populated with
[cwd, plsaTeX/Packages] in plasTeX/__init__.py . The user can then
manipulate this object after importing plasTeX. No other packages should
modify this list (especially when they are being imported), so we
wouldn't have weird race conditions as to how we should order our
imports.

The command line option will prepend to this list; more complicated
configurations require working writing python of some sort.

>Please also keep in mind that I'd like to have some plugin mechanism, as discussed in #159.

Perhaps I do not understand what you are going for here, but I don't see
what extra benefits this would provide, especially if we look for
renderers in sys.path as well.

**Comment by dalcde** [comment]
A proposal for a plugin-like mechanism. The goal is to let people package their own package implementations/renderer as modules, which can be used just by importing the module.

The layout of the module would look like
```
plasTeXplugin
├── __init__.py
├── Base
│   ├── LaTeX
│   │   └── ...
│   ├── TeX
│   │   └── ...
│   ├── Extras
│   │   └── ...
├── Packages
│   └── ...
├── Renderers
│   └── ...
```
The `__init__.py` file hooks itself to `plasTeX`, and would contain something like
```
from plasTeX.PluginManager import loadPlugin
loadPlugin(__module__)
```
It is also free to patch plasTeX if necessary.

The contents of the other folders will behave as if they were in the corresponding folders in `plasTeX`. The `Extras` folder contains definitions that should be automatically included without importing any module, but might not logically belong to `LaTeX` or `TeX`. If multiple plugins contain the same file (or `plasTeX/` contains the file), we use the one from the most recently imported plugin.

**Comment by dalcde** [comment]
This would fix #144 as well.

**Comment by PatrickMassot** [comment]
This looks good. We should also have a command line option for the main cli that would allow to load plugins.

**Comment by dalcde** [comment]
It appears that there are some similar attempts at `plasTeX/PackageResource.py`. What were your intentions with it?

**Comment by PatrickMassot** [comment]
Honestly I don't remember all details. I know I used it in my differential topology lecture notes. If you got to https://www.imo.universite-paris-saclay.fr/~pmassot/enseignement/mat553/sect0013.html and hover over the label Théorème 4.10 you'll see two icons. Clicking the graph icon will open a list of theorems used to prove that theorem. Clicking the pencil icon list exercises covering that theorem. There is also the semi-hidden page https://www.imo.universite-paris-saclay.fr/~pmassot/enseignement/mat553/coverage.html which list theorems not covered by any exercise (click on orange and red percentages to see more). 

All thoses goodies come from some version of the package below (I can't guarantee this is the latest version). At that time I was using a personal branch of plasTeX. Things should be much easier.
```python
"""
Package thmextras
Goodies for theorem environment and sample specific plasTeX package.

Options:
dep_graph: produce dependency graph

d3_url: url for the D3js library
jquery_url: url for the jquery library
dep_graph_target: dependency graph output file (relative to global output
directory)
dep_graph_tpl: template file for dependency graph, relative to the current
directory

quizz: enable quizz generation

coverage: produce coverage report

coverage_tpl: template file for coverage report, relative to the current
directory
coverage_target: coverage report output file (relative to global output
directory)
coverage_thms: list of theorem environment covered, separated by +
coverage_sectioning: coverage report section grouping

showmore: enable buttons showing or hiding proofs.

"""
import os
import string
from jinja2 import Template

from plasTeX import Command
from plasTeX.Base.LaTeX.Lists import List
from plasTeX.PackageResource import (
        PackageTemplateDir, PackageJs, PackageCss, PackageResource)

from plasTeX.Logging import getLogger
log = getLogger()

class DepGraph():
    def __init__(self):
        self.nodes = set()
        self.edges = set()

class uses(Command):
    """ \\uses{labels list} """
    args = 'labels:list:nox'

    def digest(self, tokens):
        Command.digest(self, tokens)
        node = self.parentNode
        doc = self.ownerDocument
        def update_used():
            labels_dict = doc.context.labels
            used = [labels_dict[label] for label in self.attributes['labels'] if label in labels_dict]
            node.setUserData('uses', used)
            if 'thmextras_dep_graph' in doc.userdata:
                graph = doc.userdata.get('thmextras_dep_graph')
                graph.nodes.add(node)
                graph.nodes.update(used)
                for thm in used:
                    graph.edges.add((thm, node))

        doc.postParseCallbacks.append(update_used)

class proves(Command):
    """ \proves{label} """
    args = 'label:str'

    def digest(self, tokens):
        Command.digest(self, tokens)
        node = self.parentNode
        doc = self.ownerDocument
        def update_proved():
            labels_dict = doc.context.labels
            proved = labels_dict.get(self.attributes['label'])
            if proved:
                node.setUserData('proves', proved)
                proved.userdata['proved_by'] = node
        doc.postParseCallbacks.append(update_proved)


class quizz(List):
    class qwrong(List.item):
        pass

    class qright(List.item):
        pass

    class question(Command):
        args = 'self'

        def digest(self, tokens):
            Command.digest(self, tokens)
            node = self.parentNode
            node.userdata['question'] = self

    def digest(self, tokens):
        List.digest(self, tokens)
        node = self.parentNode
        quizzes = node.userdata.get('quizzes', [])
        quizzes.append(self)
        node.userdata['quizzes'] = quizzes


class covers(Command):
    """ \covers{labels list} """
    args = 'labels:list:nox'

    def digest(self, tokens):
        Command.digest(self, tokens)
        node = self.parentNode
        doc = self.ownerDocument
        def update_covered():
            labels_dict = doc.context.labels
            covereds = [labels_dict[label] 
                    for label in self.attributes['labels'] 
                    if label in labels_dict]
            node.setUserData('covers', covereds)
            for covered in covereds:
                covered_by = covered.userdata.get('covered_by', [])
                covered.setUserData('covered_by', covered_by + [node])
        doc.postParseCallbacks.append(update_covered)


class ThmReport():
    """"""

    def __init__(self, id, caption, statement, covered_by):
        """Constructor for ThmReport"""
        self.id = id
        self.caption = caption
        self.statement = statement
        self.covered_by = covered_by

    @classmethod
    def from_thm(cls, thm):
        covered_by = thm.userdata.get('covered_by', [])
        caption = thm.caption + ' ' + thm.ref
        return cls(thm.id, caption, unicode(thm), covered_by)


class PartialReport():
    def __init__(self, title, nb_thms, nb_not_covered, thm_reports):
        self.nb_thms = nb_thms
        self.nb_not_covered = nb_not_covered
        self.coverage = 100 * (nb_thms - nb_not_covered) / nb_thms if nb_thms else 100
        self.thm_reports = thm_reports
        self.title = title
        if self.coverage == 100:
            self.status = 'ok'
        elif self.coverage > 0:
            self.status = 'partial'
        else:
            self.status = 'void'

    @classmethod
    def from_section(cls, section, thm_types):
        nb_thms = 0
        nb_not_covered = 0
        thm_reports = []
        theorems = []
        for thm_type in thm_types:
            theorems += section.getElementsByTagName(thm_type)
        for thm in theorems:
            nb_thms += 1
            thm_report = ThmReport.from_thm(thm)
            if not thm_report.covered_by:
                nb_not_covered += 1
            thm_reports.append(thm_report)
        return cls(section.fullTocEntry, nb_thms, nb_not_covered, thm_reports)


class Report():
    """A full report."""

    def __init__(self, partials):
        """Constructor for Report"""
        self.partials = partials
        self.nb_thms = sum([p.nb_thms for p in partials])
        self.nb_not_covered = sum([p.nb_not_covered for p in partials])
        self.coverage = 100 * (self.nb_thms - self.nb_not_covered) / self.nb_thms if self.nb_thms else 100



def ProcessOptions(options, document):
    """This is called when the package is loaded."""

    templatedir = PackageTemplateDir(
            renderers=['html5'],
            package='thmextras')

    document.addPackageResource(templatedir)

    jobname = document.userdata['jobname']
    outdir = document.config['files']['directory']
    outdir = string.Template(outdir).substitute({'jobname': jobname})

    links = PackageResource(
            renderers=['html5'],
            key='thmextras_usage_links',
            data='usage_links' in options)
    document.addPackageResource(links)

    if 'dep_graph' in options:
        d3_url = options.get('d3_url', 'https://d3js.org/d3.v4.min.js')
        jquery_url = options.get('jquery_url', 'http://code.jquery.com/jquery.min.js')
        title = options.get('title', 'Dependencies')
        document.userdata['thmextras_dep_graph'] = DepGraph()
        graph_target = options.get( 'dep_graph_target', 'dep_graph.html')
        
        default_template = os.path.join(os.path.dirname(__file__), 'dep_graph.j2')
        graph_template = options.get( 'dep_graph_tpl', default_template)
        try: 
            with open(graph_template) as src:
                graph_tpl = Template(src.read())
        except IOError:
            log.warning('DepGraph template read error, using default template')
            with open(default_template) as src:
                graph_tpl = Template(src.read())

        def makeDepGraph(document):
            graph = document.userdata['thmextras_dep_graph']
            graph_tpl.stream(
                    graph=graph,
                    context=document.context,
                    d3_url=d3_url,
                    jquery_url=jquery_url,
                    title=title,
                    config=document.config).dump(graph_target)
            return [graph_target]

        cb = PackageResource(
                renderers=['html5'],
                key='preCleanupCallbacks',
                data=makeDepGraph)
        css = PackageCss(
                renderers=['html5'],
                package='thmextras',
                data='dep_graph.css')
        js = PackageJs(
                renderers=['html5'],
                package='thmextras',
                data='dep_graph.js')
        document.addPackageResource([cb, css, js])

    if 'quizz' in options:
        js = PackageJs(
                renderers=['html5'],
                package='thmextras',
                data='quizz.js')
        document.addPackageResource(js)

    if 'coverage' in options:
        default_template = os.path.join(os.path.dirname(__file__), 'coverage.j2')
        coverage_template = options.get( 'coverage_tpl', default_template)
        try: 
            with open(coverage_template) as src:
                cov_tpl = Template(src.read())
        except IOError:
            log.warning('Coverage template read error, using default template')
            with open(default_template) as src:
                cov_tpl = Template(src.read())


        coverage_target = options.get( 'coverage_target', 'coverage.html')
        outfile = os.path.join(outdir, coverage_target)

        thm_types = [thm.strip() 
                for thm in options.get('coverage_thms', '').split('+')]
        section = options.get('coverage_sectioning', 'chapter')

        def makeCoverageReport(document):
            sections = document.getElementsByTagName(section)
            report = Report([PartialReport.from_section(sec, thm_types) for sec in sections])
            cov_tpl.stream(
                    report=report, 
                    config=document.config,
                    terms=document.context.terms).dump(outfile)
            return [outfile]

        cb = PackageResource(
                renderers=['html5'],
                key='preCleanupCallbacks',
                data=makeCoverageReport)
        css = PackageCss(
                renderers=['html5'],
                package='thmextras',
                data='style_coverage.css')
        js = PackageJs(
                renderers=['html5'],
                package='thmextras',
                data='coverage.js')
        document.addPackageResource([cb, css, js])

    if 'showmore' in options:
        nav = PackageResource(
                renderers=['html5'],
                key='extra-nav',
                data=[
                    {'icon': 'eye-minus',
                    'id': 'showmore-minus',
                    'class': 'showmore'},
                    {'icon': 'eye-plus',
                    'id': 'showmore-plus',
                    'class': 'showmore'}])
        css = PackageCss(
                renderers=['html5'],
                package='thmextras',
                data='showmore.css')
        js = PackageJs(
                renderers=['html5'],
                package='thmextras',
                data='showmore.js')
        js2 = PackageJs(
                renderers=['html5'],
                package='thmextras',
                data='jquery.cookie.js')
        document.addPackageResource([nav, css, js, js2])
```

**Comment by PatrickMassot** [comment]
I should probably also comment on the relation to this issue. This `PackageResource` thing is orthogonal to the issue here. They provide renderer hooks for TeX packages defined in python. But it has nothing to do with overriding package implementation or let users use their own new packages.

**Comment by PatrickMassot** [comment]
@dalcde For coordination purposes, could you tell me whether you intend to work on this issue soon?

**Comment by dalcde** [comment]
I'm not working on this at the moment. I'm currently thinking about
getting rid of ConfigManager.

**Comment by dalcde** [comment]
After some thought, I think it is better for plugins to not have global effect. Instead, a document can have a collection of plugins and the plugins only affect the particular document.

**Comment by PatrickMassot** [comment]
> After some thought, I think it is better for plugins to not have global effect. Instead, a document can have a collection of plugins and the plugins only affect the particular document.

Do you mean you want a command-line/config-file option that trigger use of each plugin you want to use with a particular document?

**Comment by dalcde** [comment]
Sure, but more importantly, if I am using plasTeX from python, I can use
it to compile different documents with different plugins.

**Comment by PatrickMassot** [comment]
@dalcde, could you review the code in https://github.com/plastex/plastex/tree/packages-plugins? I tried to work on a number a related issues there, including this one. Commits should be read one after the other, and you can add commit comments, even if there is no PR. As usual it should be better to first read modification to the Doc folder and the tests.

**Comment by dalcde** [comment]
I've had a quick look. By default, do plugins do anything other than supplying packages and renderers? (Of course, it can manually load and modify plasTeX)

**Comment by PatrickMassot** [comment]
Currently plugins can supply packages and renderers yes. Note that packages are searched in plugins before being searched in plasTeX. Hence plugins can override packages implementations.

I thought a bit about overriding more (things in `Base` or even higher in package root folder), but it gets really tricky. Also I hope it is not necessary. I don't think LaTeX packages can modify TeX that much. Do you have any use case in mind?

**Comment by dalcde** [comment]
One application I'm thinking of is to make commands and environments remember the line number (and file) they came from (see also #222). Ultimately, the goal is to streamline the process of letting people submit corrections, perhaps automatically generating a pull request from some user input.

**Comment by dalcde** [comment]
I looked through the code and it seems reasonable. I have not tested anything. I also haven't looked at the `PackageResources` part because I never grokked the original intention.

**Comment by PatrickMassot** [comment]
Thank you very much for looking through it. I'll rebase it on top of your ConfigManager work as soon as you'll fix mypy there.

**Comment by PatrickMassot** [comment]
Fixed in e9d12a2191196e28d5049d14c4b86e9d475acc31
